File: backend/app/services/auth_service.py
from fastapi import HTTPException, status
from prisma import Prisma

from app.schemas.auth_schema import LoginRequest, RegisterRequest, TokenResponse
from app.schemas.user_schema import UserResponse
from app.services.user_service import create_user, get_user_by_email
from app.utils.security import create_access_token, verify_password


async def register(db: Prisma, data: RegisterRequest) -> TokenResponse:
    existing = await get_user_by_email(db, data.email)
    if existing:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="E-mail já cadastrado.",
        )

    user = await create_user(db, data)
    token = create_access_token({"sub": user.id})

    return TokenResponse(
        access_token=token,
        user=UserResponse(
            id=user.id,
            email=user.email,
            name=user.name,
            isActive=user.isActive,
            createdAt=user.createdAt,
        ),
    )


async def login(db: Prisma, data: LoginRequest) -> TokenResponse:
    user = await get_user_by_email(db, data.email)

    if not user or not verify_password(data.password, user.password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Credenciais inválidas.",
        )

    if not user.isActive:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Conta desativada.",
        )

    token = create_access_token({"sub": user.id})

    return TokenResponse(
        access_token=token,
        user=UserResponse(
            id=user.id,
            email=user.email,
            name=user.name,
            isActive=user.isActive,
            createdAt=user.createdAt,
        ),
    )

File: backend/src/services/auth.service.ts
import { PrismaClient } from "@prisma/client";
import { createAccessToken } from "../utils/jwt";
import { verifyPassword } from "../utils/password";
import { createUser, getUserByEmail } from "./user.service";

export interface RegisterInput {
  email: string;
  password: string;
  name?: string;
}

export interface LoginInput {
  email: string;
  password: string;
}

export interface TokenResponse {
  access_token: string;
  token_type: "bearer";
  user: {
    id: string;
    email: string;
    name: string | null;
    isActive: boolean;
    createdAt: Date;
  };
}

export async function register(
  prisma: PrismaClient,
  data: RegisterInput
): Promise<TokenResponse> {
  const existing = await getUserByEmail(prisma, data.email);
  if (existing) {
    throw { statusCode: 409, message: "E-mail já cadastrado." };
  }

  const user = await createUser(prisma, data);
  const token = createAccessToken(user.id);

  return {
    access_token: token,
    token_type: "bearer",
    user: {
      id: user.id,
      email: user.email,
      name: user.name,
      isActive: user.isActive,
      createdAt: user.createdAt,
    },
  };
}

export async function login(
  prisma: PrismaClient,
  data: LoginInput
): Promise<TokenResponse> {
  const user = await getUserByEmail(prisma, data.email);

  if (!user || !(await verifyPassword(data.password, user.password))) {
    throw { statusCode: 401, message: "Credenciais inválidas." };
  }

  if (!user.isActive) {
    throw { statusCode: 403, message: "Conta desativada." };
  }

  const token = createAccessToken(user.id);

  return {
    access_token: token,
    token_type: "bearer",
    user: {
      id: user.id,
      email: user.email,
      name: user.name,
      isActive: user.isActive,
      createdAt: user.createdAt,
    },
  };
}
